src/Orleans.Transactions/InClusterTM/TransactionManager.cs

```diff
@@ -5,10 +5,99 @@
+        private const string AbortTransactionRequestTPS = "TransactionManager.AbortTransaction.TPS";
+        private const string CommitTransactionRequestTPS = "TransactionManager.CommitTransaction.TPS";
+       
+        private const string AbortedTransactionDueToDependencyTPS = "Transaction.AbortedDueToDependency.TPS";
+
+        private const string AbortedTransactionDueToMissingInfoInTransactionTableTPS =
+            "Transaction.AbortedDueToMissingInfoInTransactionTable.TPS";
+        private const string AbortedTransactionTPS = "Transaction.Aborted.TPS";
+        //pending transaction produced TPS in the monitor window
+        private const string PendingTransactionTPS = "Transaction.Pending.TPS";
+        //TPS for validated transaction in the monitor window
+        private const string ValidatedTransactionTPS = "Transaction.Validated.TPS";
+
+        internal long StartTransactionRequestCounter { get; set; }
+        internal long AbortTransactionRequestCounter { get; set; }
+        internal long CommitTransactionRequestCounter { get; set; }
+        internal long AbortedTransactionCounter { get; set; }
+        internal long AbortedTransactionDueToDependencyCounter { get; set; }
+        internal long AbortedTransactionDueToMissingInfoInTransactionTableCounter { get; set; }
+        internal long PendingTransactionCounter {get;set;}
+        internal long ValidatedTransactionCounter { get; set; }
+
+        private DateTime lastReportTime = DateTime.Now;
+        private ITelemetryProducer telemetryProducer;
+        private PeriodicAction monitor;
+        public TransactionManagerMetrics(ITelemetryProducer telemetryProducer, TimeSpan metricReportInterval)
+        {
+            this.telemetryProducer = telemetryProducer;
+            this.monitor = new PeriodicAction(metricReportInterval, this.ReportMetrics);
+        }
+
+        public void TryReportMetrics()
+        {
+            this.monitor.TryAction(DateTime.Now);
+        }
+
+        private void ResetCounters(DateTime lastReportTime)
+        {
+            this.lastReportTime = lastReportTime;
+            this.StartTransactionRequestCounter = 0;
+            this.AbortTransactionRequestCounter = 0;
+            this.CommitTransactionRequestCounter = 0;
+
+            this.AbortedTransactionDueToDependencyCounter = 0;
+            this.AbortedTransactionDueToMissingInfoInTransactionTableCounter = 0;
+            this.PendingTransactionCounter = 0;
+            this.ValidatedTransactionCounter = 0;
+            this.AbortedTransactionCounter = 0;
+        }
+
+        private void ReportMetrics()
+        {
+            if (this.telemetryProducer == null)
+                return;
+            var now = DateTime.Now;
+            var timeSinceLastReportInSeconds = Math.Max(1, (now - this.lastReportTime).TotalSeconds);
+            var startTransactionTps = StartTransactionRequestCounter / timeSinceLastReportInSeconds;
```

Comment:
Is there no way to do this without all the code? I would have thought that we have some kind of 'per time' metric. Not pushing back on it, but it seems easy to mess up with all the duplicated code.

Reply:
our internal per time metrics/statistic system has bugs in it. I prefer not use it if possible.

Reply:
And this class is in Orleans.Transaction project. And those metric/statictis system are internal to Orleans core, if I recall correctly.